### PR TITLE
[redux-form]: fix type discrepancies

### PR DESCRIPTION
**Pull request by jeppeman**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/erikras/redux-form/blob/master/src/createReduxForm.js#L290
The type in the link is what is being passed to the callbacks provided in `ConfigProps` (ex. `onChange`), note that this does not contain the injected form props, which are currently in the type definitions for the callbacks (`P & InjectedFormProps`). This issue is addressed by this PR.
https://github.com/erikras/redux-form/blob/master/src/createReduxForm.js#L1160 <-- This type has also been fixed.

### CODE REVIEW

**Comment by uniqueiniquity** [comment]
@CarsonF @aikoven @LKay @bancek @alsiola @tehbi4 @huwmartin @m-b-davis @ethanresnick @Reggino @maddijoyce @smifun @mshaaban088 @esetnik @bwlt @mrsekut
Could I please get a maintainer review? I don't have enough domain knowledge about this library to approve.

**Comment by uniqueiniquity** [comment]
@jeppeman Please ping me late next week if no maintainers review this by then.

**Comment by RyanCavanaugh** [comment]
@jeppeman you OK with this getting merged? This looks fine with me but I don't have the requisite redux knowledge. Bummer none of the other owners showed up.

**Comment by jeppeman** [comment]
@RyanCavanaugh yeah that's fine, tested it quite a bit after as well. Yup too bad. :/

**Comment by jeppeman** [comment]
@uniqueiniquity looks like we have no takers.

**Comment by uniqueiniquity** [comment]
As Ryan said, bummer. :)
Ok, I spent some time looking into your change, and I'm a little confused.
Your claim is that the `PropsWithContext` type you linked, which is getting passed to callbacks like `onChange`, does not include the injected form props, which are currently included but you have removed. However, it looks to me like that `PropsWithContext` type basically contains all of the props from `ConfigProps` (e.g., `onChange`, `onSubmit`, `destroyOnUnmount`), but also the the injected form props that you've removed (e.g., `dirty`, `submitFailed`, `invalid`). It seems like the only reason this change didn't break things is because you added the `{ [key: string]: any }` party to the `DecoratedFormProps` type, which prevented any existing references to injected form props from being marked as errors. What am I missing here?

**Comment by jeppeman** [comment]
Thanks a lot for taking the time to review this, really appreciate it.
Yes, `PropsWithContext` does share some properties with `InjectedFormProps`, but not all of them, some properties found on `InjectedFormProps` (such as handleSubmit) are not present in `PropsWithContext` and vice versa.
Currently `P & InjectedFormProps` are being passed, this is just flat out wrong, it completely omits `ConfigProps` and as previously mentioned some of the properties defined there do not exist on `PropsWithContext`.
So basically this PR makes it so that no false promises are being made about the type passed to the callbacks. 
We could expose type definitions for https://github.com/erikras/redux-form/blob/master/src/createReduxForm.js#L218 in order to make it more specific I suppose, but that changes the scope of the PR quite a bit. Would you be up for reviewing that perhaps?

File: types/redux-form/lib/reduxForm.d.ts [hunk 1]
```diff
@@ -143,22 +143,26 @@ export interface FormInstance<FormData, P, ErrorType> extends Component<P> {
...
 }
 
-export interface DecoratedComponentClass<FormData, P, ErrorType> {
-    new(props?: P, context?: any): FormInstance<FormData, P, ErrorType>;
-}
+export type DecoratedFormProps<
+  FormData = {},
+  P = {},
+  ErrorType = string
+> = P & Partial<ConfigProps<FormData, P, ErrorType>> & { [key: string]: any };
```
Comment (andrewbranch) [inline types/redux-form/lib/reduxForm.d.ts]: I think we should make whatever other changes are necessary to eliminate this `{ [key: string]: any }`, unless the type truly does have an unknowable string index signature at runtime.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

**Comment by uniqueiniquity** [comment]
@jeppeman Sorry for the delay in responding - I may not be the one who ends up reviewing the change, but i still think you should go ahead and do it as well.

**Comment by andrewbranch** [comment]
Pinging owners again for review—this seems like a fairly substantial change and it would really be better if someone familiar with the library could review. Thank you! 🙏 

@aikoven @LKay @bancek @alsiola @tehbi4 @huwmartin @m-b-davis @ethanresnick @maddijoyce @smifun @mshaaban088 @esetnik @bwlt @mrsekut

**Comment by jeppeman** [comment]
Ping :)

**Review by weswigham — APPROVED** [review]
In lieu of any responses from any type maintainers, I'm going to say that enough DT people have laid eyes on this to merge it.
